### PR TITLE
[WIP][Attention] Enable masked MHA for topk sparse attention using FA4

### PR DESCRIPTION
**Pull request by MatthewBonanni**
_**This PR is being refactored and split up into smaller PRs:**_

* https://github.com/vllm-project/vllm/pull/47327
* https://github.com/vllm-project/vllm/pull/48770

## Purpose
For pure prefills, there is a threshold of sequence length below which a masked MHA pathway is faster than sparse MQA. The [DeepSeek V3.2 paper mentions](https://arxiv.org/abs/2512.02556) this:

> Note that for short-sequence prefilling, we specially implement a masked MHA mode to
simulate DSA, which can achieve higher efficiency under short-context conditions.

This PR implements the two optimizations mentioned in https://github.com/vllm-project/vllm/issues/31473, point 3:

1. Use un-absorbed dense MHA pathway for requests with `seq_len <= topk`
2. Otherwise, use un-absorbed masked MHA pathway for requests with `min_threshold <= seq_len <= max_threshold` (here, `min_threshold = 128` and `max_threshold = 8192`)

In practice, with a typical `topk = 2048`, the routing will be as follows:
- `seq_len <= 2048`: dense MHA
- `2048 < seq_len <= 8192`: masked MHA
- `8192 < seq_len`: sparse MQA

### Correctness
```
pytest tests/v1/attention/test_sparse_mla_backends.py
```

### Performance
```
python benchmarks/attention_benchmarks/benchmark.py --config benchmarks/attention_benchmarks/configs/mla_sparse_mha_vs_mqa.yaml
```

### Correctness
Should pass in CI (V1 Attention)

### Performance

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+            )
+            attn_out = attn_out[..., : self.v_head_dim]
+            output.copy_(attn_out.flatten(start_dim=-2))
+            return
+
+        num_decodes: int = getattr(attn_metadata, "num_decodes", 0)
+        num_prefills: int = getattr(attn_metadata, "num_prefills", 0)
+        num_decode_tokens: int = getattr(attn_metadata, "num_decode_tokens", 0)
+
+        all_seq_lens = getattr(attn_metadata, "seq_lens", None)
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: nit: can we avoid using `getattr`?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    def _concat_k_nope_k_pe(
+        self, k_nope: torch.Tensor, k_pe: torch.Tensor
+    ) -> torch.Tensor:
+        k = torch.empty(
+            (*k_nope.shape[:-1], k_nope.shape[-1] + k_pe.shape[-1]),
+            dtype=k_nope.dtype,
+            device=k_nope.device,
+        )
+        k[..., : k_nope.shape[-1]] = k_nope
+        k[..., k_nope.shape[-1] :] = k_pe
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: I think `torch.ops.vllm.flashinfer_concat_mla_k(k, k_nope, k_pe)` was added after but we should still use this

File: benchmarks/attention_benchmarks/mla_runner.py [hunk 3]
```diff
@@ -778,9 +784,22 @@ def _run_single_benchmark(
...
+        has_decode = True
+        has_prefill = False
+    elif is_sparse:
+        num_decode_tokens = getattr(metadata, "num_decode_tokens", 0)
+        num_prefills = getattr(metadata, "num_prefills", 0)
+        has_decode = num_decode_tokens > 0
+        has_prefill = num_prefills > 0
+    else:
+        has_decode = getattr(metadata, "decode", None) is not None
+        has_prefill = getattr(metadata, "prefill", None) is not None
```
Comment (LucasWilkinson) [inline benchmarks/attention_benchmarks/mla_runner.py]: nit: can we avoid using `getattr`?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        prefill_max_ql: int = getattr(attn_metadata, "prefill_max_query_len", 0)
+        has_context: bool = getattr(attn_metadata, "has_context", False)
+
+        assert prefill_qsl is not None, (
+            "Metadata must provide prefill_query_start_loc for forward_mha"
+        )
+
+        num_decodes: int = getattr(attn_metadata, "num_decodes", 0)
+        num_prefills: int = getattr(attn_metadata, "num_prefills", 0)
+        num_decode_tokens: int = getattr(attn_metadata, "num_decode_tokens", 0)
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: nit: can we avoid the use of `getattr`?

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        else:
+            all_seq_lens = getattr(attn_metadata, "seq_lens", None)
+            assert all_seq_lens is not None
+            prefill_seq_lens = all_seq_lens[num_decodes : num_decodes + num_prefills]
+
+            seq_lens_t = prefill_seq_lens.to(torch.int32)
+            cu_seqlens_k = torch.zeros(
+                num_prefills + 1, dtype=torch.int32, device=device
+            )
+            torch.cumsum(seq_lens_t, dim=0, out=cu_seqlens_k[1:])
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should move this off the hotpath into the metadata

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+            total_kv_tokens = cu_seqlens_k[-1].item()
+            max_kv_len = seq_lens_t.max().item() if num_prefills > 0 else 0
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should move this off the hotpath into the metadata

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+            block_table = getattr(attn_metadata, "block_table", None)
+            assert block_table is not None
+            prefill_block_table = block_table[num_decodes : num_decodes + num_prefills]
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should move this off the hotpath into the metadata

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+            k, v = self._gather_and_decompress_all(
+                kv_c_and_k_pe_cache,
+                prefill_block_table,
+                seq_lens_t,
+                cu_seqlens_k,
+                total_kv_tokens,
+                k_scale,
+            )
+
+            prefill_qsl_cpu = prefill_qsl.cpu()
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should construct this via `cm.query_start_loc_cpu` and put it in the metadata

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+            max_q_len = (
+                max(
+                    (prefill_qsl_cpu[i + 1] - prefill_qsl_cpu[i]).item()
+                    for i in range(num_prefills)
+                )
+                if num_prefills > 0
+                else 0
+            )
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should move this off the hotpath into the metadata

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+                return_softmax_lse=True,
+                fa_version=4,
+            )
+        else:
+            from vllm.vllm_flash_attn.cute.dense_mask import (
+                dense_mask_mod,
+                dense_mask_to_block_sparse,
+            )
+
+            prefill_qsl_cpu = prefill_qsl.cpu()
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: ditto: https://github.com/vllm-project/vllm/pull/34744/changes#r3077135947

File: vllm/model_executor/layers/attention/sparse_mla_attention.py [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+            q_lens = [
+                (prefill_qsl_cpu[i + 1] - prefill_qsl_cpu[i]).item()
+                for i in range(num_prefills)
+            ]
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/sparse_mla_attention.py]: we should move this off the hotpath into the metadata; also stylistically this should just be `prefill_qsl_cpu[1:] - prefill_qsl_cpu[:-1]`

File: vllm/model_executor/layers/attention/sparse_mla_attention.py (hunk 9 above)
Comment (LucasWilkinson) [reply vllm/model_executor/layers/attention/sparse_mla_attention.py]: also stylistically this should just be `(prefill_qsl_cpu[1:] - prefill_qsl_cpu[:-1]).max()`

**Comment by MatthewBonanni** [comment]
Superseded by https://github.com/vllm-project/vllm/pull/47327 and https://github.com/vllm-project/vllm/pull/48770
